src/main/kotlin/org/ageseries/libage/sim/thermal/Thermal.kt

```diff
@@ -0,0 +1,306 @@
+        }
+    }
+
+    /**
+     * Connect two [ThermalBodies](ThermalBody), with the given connection parameters for the underlying [ThermalConnection].
+     *
+     * The connection is automatically [add]ed to the simulation, as well as returned. The object can be used to [remove] it later.
+     */
+    fun connect(a: ThermalBody<Locator>, b: ThermalBody<Locator>, params: ThermalConnectionParameters = ThermalConnectionParameters.DEFAULT): Connection<Locator> =
+        Connection(a, b, params).also {
+            add(it)
+        }
+
+    /**
+     * Remove the bodies from the simulation.
+     *
+     * If these bodies are involved in any [Connection]s, those connections will be removed as well. This can affect the flux on other bodies.
+     */
+    fun remove(vararg bodies_removed: ThermalBody<Locator>) {
+        bodies.removeAll(bodies_removed)
+        bodies_removed.forEach { body ->
+            // toList() here is just used to make a cheap copy
+            connectionMap[body].toList().forEach {
+                remove(it)
+            }
+        }
+    }
+
+    /**
+     * Remove [Connection]s from the simulation.
+     *
+     * This does not affect the membership of the underlying [ThermalBodies](ThermalBody).
+     */
+    fun remove(vararg connections_removed: Connection<Locator>) {
+        connections.removeAll(connections_removed)
+        connections_removed.forEach {
+            connectionMap[it.a].remove(it)
+            connectionMap[it.b].remove(it)
+        }
+    }
+
+    /**
+     * Run the simulation for a step.
+     *
+     * This does a discrete step of [dt] seconds on all bodies and connections in the simulation. Due to the approximations involved in the model (and the primitive integration used here), smaller step sizes are generally more stable.
+     *
+     * Flux between bodies (via [Connection]s) and with the [environment] are considered "simultaneously"--energy updates are deferred until the end.
+     */
+    fun step(dt: Double) {
+        val delta_E = mutableMapOf<ThermalBody<Locator>, Double>()
+        for (connection in connections) {
+            val transfer = connection.connection.transfer(dt)
+            delta_E[connection.a] = delta_E.getOrDefault(connection.a, 0.0) + transfer
+            delta_E[connection.b] = delta_E.getOrDefault(connection.b, 0.0) - transfer
+        }
+        for(body in bodies) {
+            val temp = environment.temperature(body.locator)
+            val cond = environment.conductance(body.locator)
+            // FIXME: integrate this better with ThermalConnection above
+            val delta_T = temp.kelvin - body.mass.temperature.kelvin
+            val overall_cond = sqrt(cond * body.surfaceArea * body.mass.material.conductance)
+            val power = overall_cond * delta_T
+            delta_E[body] = delta_E.getOrDefault(body, 0.0) + power * dt
+        }
+        for((body, delta) in delta_E.entries) {
+            body.mass.energy += delta
+        }
```

Comment:
Kotlinize.

```kotlin
connections.forEach {
    // do stuff
}
```

Reply:
blame IDEA, it was the first suggestion when I tab-completed and I was too lazy to change it, figuring nothing of value was lost :)